superset/assets/javascripts/explore/reducers/exploreReducer.js

```diff
@@ -5,77 +5,118 @@ import * as actions from '../actions/exploreActions';
-          controlPanelAlert: action.error,
-        });
+      return {
+        ...state,
+        isDatasourceMetaLoading: false,
+        controlPanelAlert: action.error,
+      };
     },
     [actions.SET_DATASOURCE]() {
-      return Object.assign({}, state, { datasource: action.datasource });
+
+      return {
+        ...state,
+        datasource: action.datasource,
+      };
     },
     [actions.FETCH_DATASOURCES_STARTED]() {
-      return Object.assign({}, state, { isDatasourcesLoading: true });
+
+      return {
+        ...state,
+        isDatasourcesLoading: true,
+      };
     },
     [actions.FETCH_DATASOURCES_SUCCEEDED]() {
-      return Object.assign({}, state, { isDatasourcesLoading: false });
+
+      return {
+        ...state,
+        isDatasourcesLoading: false,
+      };
     },
     [actions.FETCH_DATASOURCES_FAILED]() {
-      // todo(alanna) handle failure/error state
-      return Object.assign({}, state,
-        {
-          isDatasourcesLoading: false,
-          controlPanelAlert: action.error,
-        });
+
+      return {
+        ...state,
+        isDatasourcesLoading: false,
+        controlPanelAlert: action.error,
+      };
     },
     [actions.SET_DATASOURCES]() {
-      return Object.assign({}, state, { datasources: action.datasources });
+      return {
+        ...state,
+        datasources: action.datasources,
+      };
     },
     [actions.REMOVE_CONTROL_PANEL_ALERT]() {
-      return Object.assign({}, state, { controlPanelAlert: null });
+      return {
+        ...state,
+        controlPanelAlert: null,
+      };
     },
     [actions.SET_FIELD_VALUE]() {
       const controls = Object.assign({}, state.controls);
       const control = Object.assign({}, controls[action.controlName]);
       control.value = action.value;
       control.validationErrors = action.validationErrors;
       controls[action.controlName] = control;
-      const changes = { controls };
+      const changes = {
+        controls,
+      };
       if (control.renderTrigger) {
         changes.triggerRender = true;
       }
-      return Object.assign({}, state, changes);
+      return {
+        ...state,
+        ...changes,
+      };
     },
     [actions.SET_EXPLORE_CONTROLS]() {
       const controls = getControlsState(state, action.formData);
-      return Object.assign({}, state, { controls });
+      return {
+        ...state,
+        ...controls,
+      };
     },
     [actions.UPDATE_CHART_TITLE]() {
       const updatedSlice = Object.assign({}, state.slice, { slice_name: action.slice_name });
-      return Object.assign({}, state, { slice: updatedSlice });
+      return {
+        ...state,
+        slice: updatedSlice,
+      };
     },
     [actions.RESET_FIELDS]() {
       const controls = getControlsState(state, getFormDataFromControls(state.controls));
-      return Object.assign({}, state, { controls });
+      return {
+        ...state,
+        ...controls,
```

Comment:
I'm not sure if this line is right, shouldn't this be controls: just `controls,` ?

Reply:
I was following the same pattern for the `Object.assign` which is merging state and controls together. I've refactored it using the spread operator notation.
```
> a
{ foo: 'bar' }
> b
{ hugh: 'miles' }
> {...a, ...b}
{ foo: 'bar', hugh: 'miles' }
```